File: frameworks/spark/dispatcher/src/main/scala/org/apache/amaterasu/frameworks/spark/dispatcher/runners/providers/SparkSubmitScalaRunnerProvider.scala
package org.apache.amaterasu.frameworks.spark.dispatcher.runners.providers

import java.io.File

import org.apache.amaterasu.common.configuration.ClusterConfig
import org.apache.amaterasu.common.dataobjects.ActionData
import org.apache.amaterasu.common.utils.ArtifactUtil
import org.apache.amaterasu.sdk.frameworks.RunnerSetupProvider

import scala.collection.JavaConverters._

class SparkSubmitScalaRunnerProvider extends RunnerSetupProvider {

  private var conf: ClusterConfig = _
  val jarFile = new File(this.getClass.getProtectionDomain.getCodeSource.getLocation.getPath)
  val amaDist = new File(s"${new File(jarFile.getParent).getParent}/dist")
  val amalocation = new File(s"${new File(jarFile.getParent).getParent}")

  override def getCommand(jobId: String, actionData: ActionData, env: String, executorId: String, callbackAddress: String): String = {

    val util = new ArtifactUtil(List(actionData.repo).asJava, jobId)
    val classParam = if (actionData.getHasArtifact) s" --class ${actionData.entryClass}" else ""
    s"$$SPARK_HOME/bin/spark-submit $classParam ${util.getLocalArtifacts(actionData.getArtifact).get(0).getName} --deploy-mode client --jars spark-runtime-${conf.version}.jar >&1"

  }

  override def getRunnerResources: Array[String] =
    Array[String]()

  override def getActionUserResources(jobId: String, actionData: ActionData): Array[String] = {
    Array[String]()
  }

  override def getActionDependencies(jobId: String, actionData: ActionData): Array[String] = {
    val util = new ArtifactUtil(List(actionData.repo).asJava, jobId)
    conf.mode match {
      case "mesos" => util.getLocalArtifacts(actionData.getArtifact).toArray().map(x => amaDist.toPath.relativize(x.asInstanceOf[File].toPath).toString)
      case "yarn" => util.getLocalArtifacts(actionData.getArtifact).toArray().map(x => x.asInstanceOf[File].getPath)
    }
  }

  override def getHasExecutor: Boolean = false

}

object SparkSubmitScalaRunnerProvider {
  def apply(conf: ClusterConfig): SparkSubmitScalaRunnerProvider = {
    val result = new SparkSubmitScalaRunnerProvider

    result.conf = conf
    result
  }
}

File: frameworks/spark/dispatcher/src/main/kotlin/org/apache/amaterasu/frameworks/spark/dispatcher/runners/providers/SparkSubmitScalaRunnerProvider.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.amaterasu.frameworks.spark.dispatcher.runners.providers

import com.uchuhimo.konf.Config
import org.apache.amaterasu.common.configuration.ClusterConfig
import org.apache.amaterasu.common.dataobjects.ActionData
import org.apache.amaterasu.common.utils.ArtifactUtil
import org.apache.amaterasu.common.configuration.Job
import org.apache.amaterasu.sdk.frameworks.RunnerSetupProvider


class SparkSubmitScalaRunnerProvider(val conf: ClusterConfig) : RunnerSetupProvider() {

    override fun getCommand(jobId: String, actionData: ActionData, env: Config, executorId: String, callbackAddress: String): String {
        val util = ArtifactUtil(listOf(actionData.repo), jobId)
        val classParam = if (actionData.hasArtifact) " --class ${actionData.entryClass}" else ""
        val sparkProperties: Map<String, Any> = env["sparkProperties"]
        val sparkOptions: Map<String, Any> = env["sparkOptions"]
        val master = if (!sparkOptions.containsKey("master")) " --master ${env[Job.master]} " else ""

        return "\$SPARK_HOME/bin/spark-submit $classParam ${util.getLocalArtifacts(actionData.artifact).first().name} " +
                SparkCommandLineHelper.getOptions(sparkOptions) + " " +
                SparkCommandLineHelper.getProperties(sparkProperties) + " " +
                master +
                " --jars spark-runtime-${conf.version()}-all.jar >&1"
    }

    override fun getActionUserResources(jobId: String, actionData: ActionData): Array<String> = arrayOf()

    override fun getActionDependencies(jobId: String, actionData: ActionData): Array<String> = arrayOf()

    override val hasExecutor: Boolean = false

    override val runnerResources: Array<String> = arrayOf()

}